hnn_core/viz.py

```diff
@@ -64,20 +77,37 @@ def plot_hist_input(net, ax=None, show=True):
         The matplotlib figure handle.
     """
     import matplotlib.pyplot as plt
-    spikes = np.array(sum(net.spikes.times, []))
-    gids = np.array(sum(net.spikes.gids, []))
-    valid_gids = np.r_[[v for (k, v) in net.gid_dict.items()
-                        if k.startswith('evprox')]]
-    mask_evprox = np.in1d(gids, valid_gids)
-    valid_gids = np.r_[[v for (k, v) in net.gid_dict.items()
-                        if k.startswith('evdist')]]
-    mask_evdist = np.in1d(gids, valid_gids)
-    bins = np.linspace(0, net.params['tstop'], 50)
+    spike_times = np.array(sum(spikes._times, []))
+    spike_types_data = np.array(sum(spikes._types, []))
+    default_types = ['evprox', 'evdist', 'common', 'extgauss', 'extpois']
+    cell_types = ['L5_pyramidal', 'L5_basket', 'L2_pyramidal', 'L2_basket']
+    input_types = np.setdiff1d(np.unique(spike_types_data), cell_types)
+    spike_types_mask = {s_type: np.in1d(spike_types_data, s_type)
+                        for s_type in input_types}
+
+    if type(spike_types) is str:
+        spike_types = {spike_types: [spike_types]}
+    elif type(spike_types) is list:
+        spike_types = {s_type: [s_type] for s_type in spike_types}
+    elif spike_types is None:
+        spike_types = {s_type: [s_type] for s_type in default_types}
+
+    spike_mask = {s_label: np.logical_or.reduce(
+        np.logical_or.reduce([
+            [s_mask for (s_key, s_mask) in spike_types_mask.items()
+                if s_key.startswith(s_type)]
+            for s_type in s_list]))
+        for (s_label, s_list) in spike_types.items()}
```

Comment:
oh boy ...

I was going to comment that I was impressed by your clean code

can you simplify this double list comprehension into a simple nested for loop?

Reply:
Triple comprehension is not a proud moment for me...

I've unpacked this into [for loops,](https://github.com/jonescompneurolab/hnn-core/blob/b802421711cf4e8e6af3e2ad37a814575a221f65/hnn_core/viz.py#L98) but it is somewhat unweildy. I'm not totally satisfied with the variable names but hopefully it is more legible.